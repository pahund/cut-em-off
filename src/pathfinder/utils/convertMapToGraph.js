import { getNodeByCoords, initGraphAndAllowed } from './index.js';
import { N, S, W, E } from '../../directions/index.js';

export default ({ width, layers }, layer) => {
    const { graph, allowed } = initGraphAndAllowed(layers[layer].data, width);
    for (const [node, adjacent] of graph) {
        const currAllowed = allowed.get(node);
        const { row, col } = node;
        let target;
        for (const dir of currAllowed) {
            switch (dir) {
                case N:
                    target = getNodeByCoords(graph, { row: row - 1, col });
                    if (target) {
                        adjacent.set(target, 1);
                    }
                    break;
                case E:
                    target = getNodeByCoords(graph, { row, col: col + 1 });
                    if (target) {
                        adjacent.set(target, 1);
                    }
                    break;
                case S:
                    target = getNodeByCoords(graph, { row: row + 1, col });
                    if (target) {
                        adjacent.set(target, 1);
                    }
                    break;
                case W:
                    target = getNodeByCoords(graph, { row, col: col - 1 });
                    if (target) {
                        adjacent.set(target, 1);
                    }
                    break;
                default:
            }
        }
    }
    return graph;
};
