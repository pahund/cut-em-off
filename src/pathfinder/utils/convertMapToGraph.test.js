import convertMapToGraph from './convertMapToGraph';

const map = {
    width: 4,
    height: 4,
    layers: {
        main: {
            // prettier-ignore
            data: [
                20,  6, 11,  2,
                 0,  3,  0,  3,
                 1,  4,  6, 10,
                 9, 11, 10,  0
            ]
        }
    }
};

describe('When I convert a map to a graph', () => {
    describe('the result', () => {
        let result;
        beforeEach(() => (result = convertMapToGraph(map, 'main')));
        it('is correct', () => expect(result).toMatchSnapshot());
    });
});
