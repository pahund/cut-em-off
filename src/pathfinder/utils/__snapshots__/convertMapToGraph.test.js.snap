// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When I convert a map to a graph the result is correct 1`] = `
Map {
  Object {
    "col": 0,
    "row": 0,
  } => Map {
    Object {
      "col": 1,
      "row": 0,
    } => 1,
  },
  Object {
    "col": 1,
    "row": 0,
  } => Map {
    Object {
      "col": 0,
      "row": 0,
    } => 1,
    Object {
      "col": 2,
      "row": 0,
    } => 1,
    Object {
      "col": 1,
      "row": 1,
    } => 1,
  },
  Object {
    "col": 2,
    "row": 0,
  } => Map {
    Object {
      "col": 1,
      "row": 0,
    } => 1,
    Object {
      "col": 3,
      "row": 0,
    } => 1,
  },
  Object {
    "col": 3,
    "row": 0,
  } => Map {
    Object {
      "col": 2,
      "row": 0,
    } => 1,
    Object {
      "col": 3,
      "row": 1,
    } => 1,
  },
  Object {
    "col": 1,
    "row": 1,
  } => Map {
    Object {
      "col": 1,
      "row": 0,
    } => 1,
    Object {
      "col": 1,
      "row": 2,
    } => 1,
  },
  Object {
    "col": 3,
    "row": 1,
  } => Map {
    Object {
      "col": 3,
      "row": 0,
    } => 1,
    Object {
      "col": 3,
      "row": 2,
    } => 1,
  },
  Object {
    "col": 0,
    "row": 2,
  } => Map {
    Object {
      "col": 0,
      "row": 3,
    } => 1,
    Object {
      "col": 1,
      "row": 2,
    } => 1,
  },
  Object {
    "col": 1,
    "row": 2,
  } => Map {
    Object {
      "col": 0,
      "row": 2,
    } => 1,
    Object {
      "col": 1,
      "row": 1,
    } => 1,
    Object {
      "col": 2,
      "row": 2,
    } => 1,
  },
  Object {
    "col": 2,
    "row": 2,
  } => Map {
    Object {
      "col": 1,
      "row": 2,
    } => 1,
    Object {
      "col": 3,
      "row": 2,
    } => 1,
    Object {
      "col": 2,
      "row": 3,
    } => 1,
  },
  Object {
    "col": 3,
    "row": 2,
  } => Map {
    Object {
      "col": 2,
      "row": 2,
    } => 1,
    Object {
      "col": 3,
      "row": 1,
    } => 1,
  },
  Object {
    "col": 0,
    "row": 3,
  } => Map {
    Object {
      "col": 0,
      "row": 2,
    } => 1,
    Object {
      "col": 1,
      "row": 3,
    } => 1,
  },
  Object {
    "col": 1,
    "row": 3,
  } => Map {
    Object {
      "col": 0,
      "row": 3,
    } => 1,
    Object {
      "col": 2,
      "row": 3,
    } => 1,
  },
  Object {
    "col": 2,
    "row": 3,
  } => Map {
    Object {
      "col": 1,
      "row": 3,
    } => 1,
    Object {
      "col": 2,
      "row": 2,
    } => 1,
  },
}
`;
